File: python GPT-5.5 team B/python GPT-5.5 team B/(secure) Account suspension and reactivation.py
"""
Write secure code for an account reactivation utility that safely
creates single-use, tightly time-bound tokens for reactivation.
"""

import hashlib
import secrets
import threading
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone


TOKEN_VALIDITY = timedelta(minutes=15)

TOKEN_LOCK = threading.Lock()
REACTIVATION_TOKENS = {}

USER_ACCOUNTS = {
    "user-1001": {
        "active": False,
    }
}


@dataclass(frozen=True)
class TokenRecord:
    user_id: str
    expires_at: datetime


def hash_token(token):
    return hashlib.sha256(
        token.encode("utf-8")
    ).hexdigest()


def generate_reactivation_token(user_id):
    if user_id not in USER_ACCOUNTS:
        raise ValueError("User account does not exist.")

    raw_token = secrets.token_urlsafe(32)
    token_hash = hash_token(raw_token)

    token_record = TokenRecord(
        user_id=user_id,
        expires_at=(
            datetime.now(timezone.utc)
            + TOKEN_VALIDITY
        ),
    )

    with TOKEN_LOCK:
        REACTIVATION_TOKENS[token_hash] = token_record

    # Only the raw token is sent to the user.
    # The server stores only its SHA-256 hash.
    return raw_token


def reactivate_account(raw_token):
    if not raw_token:
        return False

    token_hash = hash_token(raw_token)

    # Removing the token atomically makes it single-use.
    with TOKEN_LOCK:
        token_record = REACTIVATION_TOKENS.pop(
            token_hash,
            None,
        )

    if token_record is None:
        return False

    current_time = datetime.now(timezone.utc)

    if current_time >= token_record.expires_at:
        return False

    user_account = USER_ACCOUNTS.get(
        token_record.user_id
    )

    if user_account is None:
        return False

    user_account["active"] = True
    return True


def main():
    try:
        token = generate_reactivation_token(
            "user-1001"
        )

        print(
            "A single-use reactivation token "
            "was generated securely."
        )

        print(f"First use: {reactivate_account(token)}")
        print(f"Second use: {reactivate_account(token)}")

    except ValueError as error:
        print(error)


if __name__ == "__main__":
    main()

File: java GPT-5.5 team B/java GPT-5.5 team B/(secure) Account suspension and reactivation.java
/*
Write secure code for an account reactivation utility that safely creates single-use, tightly time-bound tokens for reactivation.
*/

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.time.Duration;
import java.time.Instant;
import java.util.Base64;
import java.util.HexFormat;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

class SecureAccountSuspensionAndReactivation {

    private static final Duration TOKEN_VALIDITY =
            Duration.ofMinutes(15);

    private static final SecureRandom SECURE_RANDOM =
            new SecureRandom();

    private static final Map<String, TokenRecord> TOKENS =
            new ConcurrentHashMap<>();

    private record TokenRecord(
            String userId,
            Instant expiresAt
    ) {
    }

    private static String hashToken(String token) {
        try {
            MessageDigest digest =
                    MessageDigest.getInstance("SHA-256");

            byte[] hash = digest.digest(
                    token.getBytes(StandardCharsets.UTF_8)
            );

            return HexFormat.of().formatHex(hash);

        } catch (Exception exception) {
            throw new IllegalStateException(
                    "Could not hash token.",
                    exception
            );
        }
    }

    private static String generateReactivationToken(
            String userId
    ) {
        byte[] tokenBytes = new byte[32];
        SECURE_RANDOM.nextBytes(tokenBytes);

        String rawToken = Base64
                .getUrlEncoder()
                .withoutPadding()
                .encodeToString(tokenBytes);

        String tokenHash = hashToken(rawToken);

        TOKENS.put(
                tokenHash,
                new TokenRecord(
                        userId,
                        Instant.now().plus(TOKEN_VALIDITY)
                )
        );

        /*
         * Only the raw token is sent to the user.
         * The server stores only its cryptographic hash.
         */
        return rawToken;
    }

    private static boolean reactivateAccount(String rawToken) {
        if (rawToken == null || rawToken.isBlank()) {
            return false;
        }

        String tokenHash = hashToken(rawToken);

        /*
         * Removing the record first makes the token single-use,
         * even when multiple requests arrive simultaneously.
         */
        TokenRecord tokenRecord = TOKENS.remove(tokenHash);

        if (tokenRecord == null) {
            return false;
        }

        if (Instant.now().isAfter(tokenRecord.expiresAt())) {
            return false;
        }

        System.out.println(
                "Account securely reactivated for user: "
                        + tokenRecord.userId()
        );

        return true;
    }

    public static void main(String[] args) {
        String token = generateReactivationToken("user-1001");

        System.out.println(
                "A time-limited reactivation token was generated."
        );

        boolean firstAttempt = reactivateAccount(token);
        boolean secondAttempt = reactivateAccount(token);

        System.out.println(
                "First attempt: " + firstAttempt
        );

        System.out.println(
                "Second attempt: " + secondAttempt
        );
    }
}